webapi/middleware.go

```diff
@@ -305,58 +305,93 @@ func vspAuth() gin.HandlerFunc {
 			return
 		}
 
-		// If the ticket was found in the database, we already know its
-		// commitment address. Otherwise we need to get it from the chain.
-		var commitmentAddress string
-		if ticketFound {
-			commitmentAddress = ticket.CommitmentAddress
-		} else {
-			dcrdClient := c.MustGet("DcrdClient").(*rpc.DcrdRPC)
+		altSigAddr, err := db.AltSigAddr(hash)
+		if err != nil {
+			log.Errorf("%s: db.AltSigAddr (ticketHash=%s): %v", funcName, hash, err)
+			sendError(errInternalError, c)
+			return
+		}
 
-			resp, err := dcrdClient.GetRawTransaction(hash)
-			if err != nil {
-				log.Errorf("%s: dcrd.GetRawTransaction for ticket failed (ticketHash=%s): %v", funcName, hash, err)
-				sendError(errInternalError, c)
-				return
+		// validate validates the signature in the request as signing
+		// the request body with private key corresponding to sAddr.
+		// Verification failure results in the client getting notified
+		// of the cause. Returns whether verification succeeded.
+		validate := func(kt bool) {
+			t := &ticket
+			validateSig := func(addr string) bool {
+				// Validate request signature to ensure ticket ownership.
+				err = validateSignature(reqBytes, addr, c)
+				if err != nil {
+					log.Warnf("%s: Bad signature (clientIP=%s, ticketHash=%s): %v", funcName, c.ClientIP(), hash, err)
+					return false
+				}
+				log.Tracef("%s: Authed user signature (clientIP=%s, ticketHash=%s)", funcName, c.ClientIP(), hash)
+				return true
 			}
 
-			msgTx, err := decodeTransaction(resp.Hex)
-			if err != nil {
-				log.Errorf("%s: Failed to decode ticket hex (ticketHash=%s): %v", funcName, ticket.Hash, err)
-				sendError(errInternalError, c)
-				return
+			// If an alternate signature address exists and was used
+			// in the message signature, it is the signing address.
+			// Fallback to commitment signature verification if there
+			// is no alternate or it failed.
+			var sAddr string
+			if altSigAddr != "" && validateSig(altSigAddr) {
+				sAddr = altSigAddr
+			} else {
+				if !validateSig(t.CommitmentAddress) {
+					sendError(errBadSignature, c)
+					return
+				}
+				sAddr = t.CommitmentAddress
 			}
 
-			err = isValidTicket(msgTx)
-			if err != nil {
-				log.Warnf("%s: Invalid ticket (clientIP=%s, ticketHash=%s): %v", funcName, c.ClientIP(), hash, err)
-				sendError(errInvalidTicket, c)
-				return
-			}
+			// Add ticket information to context so downstream handlers don't need
+			// to access the db for it.
+			c.Set("Ticket", *t)
+			c.Set("KnownTicket", kt)
+			c.Set("CommitmentAddress", t.CommitmentAddress)
+			c.Set("SigningAddress", sAddr)
```

Comment:
I dont understand why this is being added into the context. It's only used in `setaltsig.go`, and I think the code there can just extract it from `SetAltSigRequest`.

If we are able to remove `aAddr`, that simplifies the code above a bit as well.

Reply:
I think you are correct. Will try.

Reply:
Yeah, now that we can only set once and not clear, it can only ever be the commitment address.

Reply:
9aebd65, 0175871, and c709561